src/input/tablet.c

```diff
@@ -63,27 +66,163 @@ adjust_for_rotation(enum rotation rotation, double *x, double *y)
+		if (!tablet->tablet_v2_tool) {
+			wlr_log(WLR_DEBUG, "  got axis event on surface without tablet_v2_tool");
+		}
+		if (tablet->tablet_v2_tool && tablet->tablet_v2_tool->wlr_tool != ev->tool) {
+			wlr_log(WLR_DEBUG, "  got axis event on surface with wrong tablet_v2_tool");
+		}
+	}
+
+	if (surface && tablet->tablet_v2_tool && tablet->tablet_v2_tool->wlr_tool == ev->tool) {
+		if (ev->updated_axes & (WLR_TABLET_TOOL_AXIS_X | WLR_TABLET_TOOL_AXIS_Y)) {
+			// Visualize where we are and what we are
+			wlr_cursor_warp_absolute(tablet->seat->cursor, tablet->base, x, y);
+			cursor_set(tablet->seat, LAB_CURSOR_PENCIL);
+
+			wlr_tablet_v2_tablet_tool_notify_proximity_in(tablet->tablet_v2_tool,
+				tablet->tablet_v2, surface);
+			wlr_tablet_v2_tablet_tool_notify_motion(tablet->tablet_v2_tool, nx, ny);
+		}
+		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_DISTANCE) {
+			wlr_tablet_v2_tablet_tool_notify_distance(tablet->tablet_v2_tool,
+				ev->distance);
+		}
+		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_PRESSURE) {
+			wlr_tablet_v2_tablet_tool_notify_pressure(tablet->tablet_v2_tool,
+				ev->pressure);
+		}
+		if (ev->updated_axes
+				& (WLR_TABLET_TOOL_AXIS_TILT_X | WLR_TABLET_TOOL_AXIS_TILT_Y)) {
+			wlr_tablet_v2_tablet_tool_notify_tilt(tablet->tablet_v2_tool, nx, ny);
+		}
+		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_ROTATION) {
+			wlr_tablet_v2_tablet_tool_notify_rotation(tablet->tablet_v2_tool,
+				ev->rotation);
+		}
+		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_SLIDER) {
+			wlr_tablet_v2_tablet_tool_notify_rotation(tablet->tablet_v2_tool,
+				ev->slider);
+		}
+		if (ev->updated_axes & WLR_TABLET_TOOL_AXIS_WHEEL) {
+			wlr_tablet_v2_tablet_tool_notify_wheel(tablet->tablet_v2_tool,
+				ev->wheel_delta, 0);
+		}
+	} else {
+		if (ev->updated_axes & (WLR_TABLET_TOOL_AXIS_X | WLR_TABLET_TOOL_AXIS_Y)) {
+			if (tablet->tablet_v2_tool) {
+				wlr_tablet_v2_tablet_tool_notify_proximity_out(
+					tablet->tablet_v2_tool);
+			}
+
+			cursor_emulate_move_absolute(tablet->seat, &ev->tablet->base,
+				x, y, ev->time_msec);
+		}
+	}
+}
+
+static void
+handle_proximity(struct wl_listener *listener, void *data)
+{
+	wlr_log(WLR_DEBUG, "Got proximity event");
+	struct wlr_tablet_tool_proximity_event *ev = data;
+	struct drawing_tablet *tablet = ev->tablet->data;
+
+	tablet->x = ev->x;
+	tablet->y = ev->y;
+
+	double x = ev->x;
+	double y = ev->y;
+	adjust_for_tablet_area(tablet->tablet->width_mm, tablet->tablet->height_mm,
+		rc.tablet.box, &x, &y);
+	adjust_for_rotation(rc.tablet.rotation, &x, &y);
+
+	double lx, ly;
+	struct wlr_surface *surface = tablet_get_coords(tablet, x, y, &lx, &ly);
+
+	if (surface) {
+		wlr_log(WLR_DEBUG, "  got proximity event on surface");
+		// TODO: This seems just plain wrong and causes crashes
+		if (!tablet->tablet_v2_tool || tablet->tablet_v2_tool->wlr_tool != ev->tool) {
+			// Unfortunately I found `wlr_tool` only in the events,
+			// so use this for creating a `wlr_tablet_v2_tablet_tool`
```

Comment:
`struct wlr_tablet_pad` appears to have a signal for attaching a tool (`attach_tablet`).

Reply:
I looked a bit into wlroots and the issue is that wlroots destroys the tool on proximity_out when it can't be uniquely identified. So using the proximity_in signal to create the tool wrapper seems like the correct solution.